File: sReports/sReportsV2/wwwroot/js/sReports/episodeOfCare/task.js
var taskColumnName;
var taskSwitchCount = 0;
var taskIsAscending = null;

function showTaskModal(event, id, readOnly = false, reloadTaskTable = false) {
    event.stopPropagation();
    let taskId = id ? id : 0;
    setActiveClass(taskId);

    $.ajax({
        type: 'GET',
        url: `/Task/ShowTaskModal?taskId=${taskId}&isReadOnlyViewMode=${readOnly}`,
        success: function (data) {
            $('#addTaskModal').html(data);
            $('#addTaskModal').modal('show');
            document.getElementById('reloadTaskTable').value = reloadTaskTable;
            if (taskId != 0)
                reloadCodeSetChildren("taskType");
        },
        error: function (xhr, ajaxOptions, thrownError) {
            handleResponseError(xhr, true);
        }
    });

    removeActiveClassOnCloseModal();
}

function submitTaskForm() {
    updateDisabledOptions(false);
    $("#newTaskForm").validate();

    if ($("#newTaskForm").valid()) {
        var request = {};
        var encounterId = $("#activeEncounter").val();
        request['TaskId'] = $("#taskId").val();
        request['PatientId'] = $("#pId").val();
        request['EncounterId'] = encounterId;
        request['TaskTypeCD'] = $("#taskType").val();
        request['TaskStatusCD'] = $("#taskStatus").val();
        request['TaskPriorityCD'] = $("#taskPriority").val();
        request['TaskClassCD'] = $("#taskClass").val();
        request['TaskDescription'] = $("#taskDescription").val();
        request['TaskEntityId'] = $("#taskEntityId").val();
        request['TaskStartDateTime'] = calculateDateTimeWithOffset("#taskStartDateTime");
        request['TaskEndDateTime'] = calculateDateTimeWithOffset("#taskEndDateTime");
        request['ScheduledDateTime'] = calculateDateTimeWithOffset("#scheduledDateTime");
  
        $.ajax({
            type: 'POST',
            url: `/Task/Create`,
            data: request,
            success: function (data, jqXHR) {
                $('#addTaskModal').modal('hide');
                if (document.getElementById('reloadTaskTable').value === "true")
                    reloadTasks(null, null, true);
                else
                    showDetails(encounterId);
                toastr.success("Success");
            },
            error: function (xhr, thrownError) {
                handleResponseError(xhr);
            }
        });
    }
    return false;
}

function setActiveClass(taskId) {
    var taskActive = $(".blue-pencil-task").attr("data-task-id");
    if (taskActive != taskId)
        $(".edit").removeClass("blue-pencil-task");
    $(`#edit-${taskId}`).toggleClass("blue-pencil-task");
}

function removeActiveClassOnCloseModal() {
    $('#addTaskModal').on('hidden.bs.modal', function () {
        $('.blue-pencil-task').removeClass('blue-pencil-task');
    });
}

function showTasksContent(taskColumnName, taskIsAscending, isFilter) {
    $("#taskTab").addClass("code-active-tab");
    $("#eocTab").addClass("remove-eoc-tab");
    $("#eocTab").removeClass("code-active-tab");
    $("#encounterTab").removeClass("code-active-tab");
    reloadTasks(taskColumnName, taskIsAscending, isFilter);
}

function reloadTasks(taskColumnName, taskIsAscending, isFilter) {
    var request = {};
    request['PatientId'] = $('#patientId').val();
    request['IsAscending'] = taskIsAscending;
    request['ColumnName'] = taskColumnName;
    request['TaskStatus'] = $('#TaskStatusTemp').val();

    $.ajax({
        type: 'GET',
        url: '/Task/ReloadTable',
        data: request,
        success: function (data) {
            if (!isFilter)
                showTaskFilterGroup();
            $("#eocContainer").html(data);
            addTaskSortArrows();
            setTableMaxHeight("taskTable", "taskTableContent");
        },
        error: function (xhr, thrownError) {
            handleResponseError(xhr);
        }
    });
}

function sortTaskTable(column) {
    if (taskSwitchCount == 0) {
        if (taskColumnName == column)
            taskIsAscending = checkIfAsc(taskIsAscending);
        else
            taskIsAscending = true;
        taskSwitchCount++;
    }
    else {
        if (taskColumnName != column)
            taskIsAscending = true;
        else
            taskIsAscending = checkIfAsc(taskIsAscending);
        taskSwitchCount--;
    }
    taskColumnName = column;

    showTasksContent(taskColumnName, taskIsAscending, true);
}

function addTaskSortArrows() {
    var element = document.getElementById(taskColumnName);
    if (element != null) {
        element.classList.remove("sort-arrow");
        if (taskIsAscending) {
            element.classList.remove("sort-arrow-desc");
            element.classList.add("sort-arrow-asc");
        }
        else {
            element.classList.remove("sort-arrow-asc");
            element.classList.add("sort-arrow-desc");
        }
    }
}

function showTaskFilterGroup() {
    $.ajax({
        type: "GET",
        url: '/Task/ShowTaskFilterGroup',
        success: function (data) {
            $("#filterGroup").html(data);
            setTableMaxHeight("taskTable", "taskTableContent");
        },
        error: function (xhr, thrownError) {
            handleResponseError(xhr);
        }
    });
}

function taskFilter() {
    pushState();
    showTasksContent(null, null, true);
}

function clickedTaskRow(e, id) {
    if (!$(e.target).hasClass('dropdown-button') && !$(e.target).hasClass('fa-bars') && !$(e.target).hasClass('dropdown-item') && !$(e.target).hasClass('dots') && !$(e.target).hasClass('table-more')) {
        showTaskModal(event, id, true, true);
    }
}

File: sReports/sReportsV2.Domain.Sql/Entities/TaskEntry/Task.cs
using System.ComponentModel.DataAnnotations.Schema;
using System.ComponentModel.DataAnnotations;
using sReportsV2.Domain.Sql.Entities.User;
using System.Collections.Generic;
using System.Linq;
using sReportsV2.Common.Enums;
using sReportsV2.Domain.Sql.Entities.Common;
using System;
using sReportsV2.Common.Extensions;

namespace sReportsV2.Domain.Sql.Entities.TaskEntry
{
    public class Task : EntitiesBase.Entity
    {
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        [Key]
        [Column("TaskId")]
        public int TaskId { get; set; }
        public int PatientId { get; set; }
        [ForeignKey("PatientId")]
        public Patient.Patient Patient { get; set; }
        public int EncounterId { get; set; }
        [ForeignKey("EncounterId")]
        public Encounter.Encounter Encounter { get; set; }
        public int TaskTypeCD { get; set; }
        [ForeignKey("TaskTypeCD")]
        public Code TaskType { get; set; }
        public int TaskStatusCD { get; set; }
        [ForeignKey("TaskStatusCD")]
        public Code TaskStatus { get; set; }
        public int? TaskPriorityCD { get; set; }
        [ForeignKey("TaskPriorityCD")]
        public Code TaskPriority { get; set; }
        public int? TaskClassCD { get; set; }
        [ForeignKey("TaskClassCD")]
        public Code TaskClass { get; set; }
        public string TaskDescription { get; set; }
        public string TaskEntityId { get; set; }
        public DateTimeOffset TaskStartDateTime { get; set; }
        public DateTimeOffset? TaskEndDateTime { get; set; }
        public DateTimeOffset? ScheduledDateTime { get; set; }
        public int? TaskDocumentId { get; set; }
        [ForeignKey("TaskDocumentId")]
        public TaskDocument TaskDocument { get; set; }

        public void Copy(Task task)
        {
            this.TaskTypeCD = task.TaskTypeCD;
            this.TaskStatusCD = task.TaskStatusCD;
            this.TaskPriorityCD = task.TaskPriorityCD;
            this.TaskClassCD = task.TaskClassCD;
            this.TaskDescription = task.TaskDescription;
            this.TaskEntityId = task.TaskEntityId;
            this.TaskStartDateTime = task.TaskStartDateTime;
            this.TaskEndDateTime = task.TaskEndDateTime;
            this.ScheduledDateTime = task.ScheduledDateTime;
            this.TaskDocumentId = task.TaskDocumentId;
        }
    }
}
